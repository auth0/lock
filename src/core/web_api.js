import auth0 from 'auth0-js';
import Auth0LegacyAPIClient from './web_api/legacy_api'
import Auth0APIClient from './web_api/p2_api'

class Auth0WebAPI {
  constructor() {
    this.clients = {};
  }

  setupClient(lockID, clientID, domain, opts) {

    const hostedLoginPage = window.location.host === domain;
    // when it is used on on the hosted login page, it shouldn't use popup mode
    opts.redirect = hostedLoginPage ? true : opts.redirect;

    opts.oidcConformant = opts.oidcConformant || false;

    // for cordova and electron we should force popup without SSO so it uses
    // /ro or /oauth/token for DB connections
    if (window && (!!window.cordova || !!window.electron)) {
      opts.redirect = false;
      opts.sso = false;
    }

    // when it is used on on the hosted login page, it should use the legacy mode
    // (usernamepassword/login) in order to continue the transaction after authentication
    if (hostedLoginPage || !opts.oidcConformant) {
      this.clients[lockID] = new Auth0LegacyAPIClient(clientID, domain, opts);
    } else {
      this.clients[lockID] = new Auth0APIClient(lockID, clientID, domain, opts);
    }
  }

  logIn(lockID, options, authParams, cb) {
    this.clients[lockID].logIn(options, authParams, cb);
  }

  signOut(lockID, query) {
    this.clients[lockID].logout(query);
  }

  signUp(lockID, options, cb) {
    this.clients[lockID].signUp(options, cb);
  }

  resetPassword(lockID, options, cb) {
    this.clients[lockID].resetPassword(options, cb);
  }

  startPasswordless(lockID, options, cb) {
    this.clients[lockID].startPasswordless(options, cb);
  }

  parseHash(lockID, hash = '', cb) {
    return this.clients[lockID].parseHash(hash, cb);
  }

  getUserInfo(lockID, token, callback) {
    return this.clients[lockID].getUserInfo(token, callback);
  }

  getProfile(lockID, token, callback) {
    return this.clients[lockID].getProfile(token, callback);
  }

  getSSOData(lockID, ...args) {
    return this.clients[lockID].getSSOData(...args);
  }

  getUserCountry(lockID, cb) {
    return this.clients[lockID].getUserCountry((err, data) => cb(err, data && data.countryCode));
  }
}

export default new Auth0WebAPI();
