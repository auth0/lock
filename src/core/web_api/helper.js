export function normalizeError(error, domain) {
  if (!error) {
    return error;
  }

  // TODO: clean this mess, the first checks are for social/popup,
  // then we have some stuff for passwordless and the latter is for
  // db.

  // TODO: the following checks were copied from https://github.com/auth0/lock/blob/0a5abf1957c9bb746b0710b274d0feed9b399958/index.js#L1263-L1288
  // Some of the checks are missing because I couldn't reproduce them and I'm
  // affraid they'll break existent functionality if add them.
  // We need a better errror handling story in auth0.js.

  if (error.status === 'User closed the popup window') {
    // {
    //   status: "User closed the popup window",
    //   name: undefined,
    //   code: undefined,
    //   details: {
    //     description: "server error",
    //     code: undefined
    //   }
    // }
    return {
      code: 'lock.popup_closed',
      error: 'lock.popup_closed',
      description: 'Popup window closed.'
    };
  }

  if (error.code === 'unauthorized') {
    // Custom rule error
    //
    // {
    //   "code": "unauthorized",
    //   "details": {
    //     "code": "unauthorized",
    //     "error_description": "user is blocked",
    //     "error": "unauthorized"
    //   },
    //   "name": "unauthorized",
    //   "status": 401
    // }

    // Default "user is blocked" rule error
    //
    // {
    //   "code": "unauthorized",
    //   "details": {
    //     "code": "unauthorized",
    //     "error_description": "user is blocked",
    //     "error": "unauthorized"
    //   },
    //   "name": "unauthorized",
    //   "status": 401
    // }

    // Social cancel permissions.
    //
    // {
    //   code: "unauthorized",
    //   details: {
    //     code: "unauthorized"
    //     error: "unauthorized"
    //     error_description: "access_denied"
    //   },
    //   name: "unauthorized"
    //   status: 401
    // }

    // Social cancel permissions or unknown error
    if (!error.description || error.description === 'access_denied') {
      return {
        code: 'lock.unauthorized',
        error: 'lock.unauthorized',
        description: error.description || 'Permissions were not granted.'
      };
    }

    // Special case for custom rule error
    if (error.description === 'user is blocked') {
      return {
        code: 'blocked_user',
        error: 'blocked_user',
        description: error.description
      };
    }

    // Custom Rule error
    return {
      code: 'rule_error',
      error: 'rule_error',
      description: error.description
    };
  }
  if (
    window.location.host !== domain &&
    (error.error === 'access_denied' || error.code === 'access_denied')
  ) {
    return {
      code: 'invalid_user_password',
      error: 'invalid_user_password',
      description: error.description
    };
  }

  const result = {
    error: error.code ? error.code : error.statusCode || error.error,
    description: error.description || error.code
  };

  // result is used for passwordless and error for database.
  return result.error === undefined && result.description === undefined ? error : result;
}

export function loginCallback(redirect, domain, cb) {
  return redirect
    ? error => cb(normalizeError(error, domain))
    : (error, result) => cb(normalizeError(error, domain), result);
}

export function normalizeAuthParams({ popup, ...authParams }) {
  return authParams;
}

export function webAuthOverrides({ __tenant, __token_issuer, __jwks_uri } = {}) {
  if (__tenant || __token_issuer || __jwks_uri) {
    return {
      __tenant,
      __token_issuer,
      __jwks_uri
    };
  }
  return null;
}

export function trimAuthParams(params = {}) {
  const { ...p } = params;
  ['username', 'email', 'phoneNumber', 'mfa_code'].forEach(k => {
    if (typeof p[k] === 'string') {
      p[k] = p[k].trim();
    }
  });
  return p;
}
