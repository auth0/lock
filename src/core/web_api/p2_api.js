import auth0 from 'auth0-js';
import CordovaAuth0Plugin from 'auth0-js/dist/cordova-auth0-plugin.min.js';
import {
  normalizeError,
  loginCallback,
  normalizeAuthParams,
  webAuthOverrides,
  trimAuthParams
} from './helper';

class Auth0APIClient {
  constructor(lockID, clientID, domain, opts) {
    this.lockID = lockID;
    this.client = null;
    this.authOpt = null;
    this.domain = domain;
    this.isUniversalLogin = window.location.host === domain;
    this._enableIdPInitiatedLogin = !!(opts._enableIdPInitiatedLogin || opts._enableImpersonation);

    const default_telemetry = {
      name: 'lock.js',
      version: __VERSION__,
      lib_version: auth0.version
    };

    var state = opts.state;
    if (opts.params && opts.params.state) {
      state = opts.params.state;
    }

    var nonce = opts.nonce;
    if (opts.params && opts.params.nonce) {
      nonce = opts.params.nonce;
    }

    const scope = opts.params && opts.params.scope;

    this.client = new auth0.WebAuth({
      clientID: clientID,
      domain: domain,
      audience: opts.audience,
      redirectUri: opts.redirectUrl,
      responseMode: opts.responseMode,
      responseType: opts.responseType,
      leeway: opts.leeway || 1,
      plugins: opts.plugins || [new CordovaAuth0Plugin()],
      overrides: webAuthOverrides(opts.overrides),
      _sendTelemetry: opts._sendTelemetry === false ? false : true,
      _telemetryInfo: opts._telemetryInfo || default_telemetry,
      state,
      nonce,
      scope
    });

    this.authOpt = {
      popup: !opts.redirect,
      popupOptions: opts.popupOptions,
      nonce,
      state,
      scope
    };
    if (this.isUniversalLogin && opts.sso !== undefined) {
      this.authOpt.sso = opts.sso;
    }
  }

  logIn(options, authParams, cb) {
    // TODO: for passwordless only, try to clean in auth0.js
    // client._shouldRedirect = redirect || responseType === "code" || !!redirectUrl;
    const f = loginCallback(false, this.domain, cb);
    const loginOptions = trimAuthParams(
      normalizeAuthParams({
        ...options,
        ...this.authOpt,
        ...authParams
      })
    );

    if (options.login_hint) {
      loginOptions.login_hint = options.login_hint;
    }

    if (!options.username && !options.email) {
      if (this.authOpt.popup) {
        this.client.popup.authorize(
          {
            ...loginOptions,
            owp: true
          },
          f
        );
      } else {
        this.client.authorize(loginOptions, f);
      }
    } else if (this.authOpt.popup) {
      this.client.popup.loginWithCredentials(loginOptions, f);
    } else {
      loginOptions.realm = options.connection;
      this.client.login(loginOptions, f);
    }
  }

  logout(query) {
    this.client.logout(query);
  }

  signUp(options, cb) {
    delete options.autoLogin;
    this.client.signup(trimAuthParams(options), (err, result) => cb(err, result));
  }

  resetPassword(options, cb) {
    this.client.changePassword(trimAuthParams(options), cb);
  }

  passwordlessStart(options, cb) {
    this.client.passwordlessStart(trimAuthParams(options), err => cb(normalizeError(err)));
  }

  passwordlessVerify(options, cb) {
    const verifyOptions = {
      ...options,
      popup: this.authOpt.popup
    };
    this.client.passwordlessLogin(verifyOptions, (err, result) => cb(normalizeError(err), result));
  }

  parseHash(hash = '', cb) {
    return this.client.parseHash(
      {
        __enableIdPInitiatedLogin: this._enableIdPInitiatedLogin,
        hash,
        nonce: this.authOpt.nonce,
        state: this.authOpt.state
      },
      cb
    );
  }

  getUserInfo(token, callback) {
    return this.client.client.userInfo(token, callback);
  }

  getProfile(token, callback) {
    this.getUserInfo(token, callback);
  }

  getSSOData(...params) {
    return this.client.client.getSSOData(...params);
  }

  getUserCountry(cb) {
    return this.client.client.getUserCountry(cb);
  }

  checkSession(options, cb) {
    return this.client.checkSession(options, cb);
  }
}

export default Auth0APIClient;
