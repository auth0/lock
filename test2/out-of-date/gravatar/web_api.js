import expect from 'expect.js'
import { spy, stub } from 'sinon';
import * as api from '../../src/gravatar/web_api';
import jsonp from '../../src/utils/jsonp_utils';

const email = "someone@auth0.com";
const emailMD5 = "b91fa14e9ce922cc2fdedb2f84dba3a5";
const invalidEmail = "invalidEmail";
let fail, success;

describe("fetching image", function() {
  let img;

  // TODO: stub preload.img and add tests for preload.
  function getCallback(name) {
    for (var i = 0; i < img.addEventListener.callCount; i++) {
      if (img.addEventListener.getCall(i).args[0] == name) {
        return img.addEventListener.getCall(i);
      }
    }
  }

  function invokeCallback(name) {
    const args = [name === "error" ? {} : undefined];
    return getCallback(name).args[1].apply(undefined, args);
  }

  beforeEach(function() {
    img = { addEventListener: spy() };
    stub(global.document, "createElement").returns(img);
    success = spy();
    fail = spy();
  });

  afterEach(function() {
    global.document.createElement.restore();
  });

  describe("for an invalid email", function() {
    beforeEach(function() {
      api.img(invalidEmail, success, fail);
    });

    it("invokes the 'fail' callback", function() {
      expect(fail.calledOnce).to.be(true);
      expect(fail.calledWithExactly(invalidEmail)).to.be(true);

    });
  });

  describe("for a valid email", function() {
    beforeEach(function() {
      api.img(email, success, fail);
    });

    it("constructs the expected URL", function() {
      const url = `https://secure.gravatar.com/avatar/${emailMD5}?d=404`;
      expect(img.src).to.be(url);
    });

    it("setups a 'load' and a 'error' callback", function() {
      expect(img.addEventListener.calledTwice).to.be(true);
      expect(img.addEventListener.withArgs("load").calledOnce).to.be(true);
      expect(img.addEventListener.withArgs("error").calledOnce).to.be(true);
    });

    describe("with an associated Gravatar image", function() {
      beforeEach(function() {
        invokeCallback("load");
      });

      it("invokes the 'success' callback with `email` and `img`", function() {
        expect(success.calledOnce).to.be(true);
        expect(success.calledWithExactly(email, img)).to.be(true)
      });
    });

    describe("without an associated Gravatar image", function() {
      beforeEach(function() {
        invokeCallback("error");
      });

      it("invokes the 'fail' callback with `email`", function() {
        expect(fail.calledOnce).to.be(true);
        expect(fail.calledWithExactly(email)).to.be(true)
      });
    });
  });
});

describe("fetching profile", function() {
  const email = "someone@auth0.com";
  const emailMD5 = "b91fa14e9ce922cc2fdedb2f84dba3a5";
  const invalidEmail = "invalidEmail";

  function invokeCallback(err, obj) {
    return jsonp.get.lastCall.args[1].call(undefined, err, obj);
  }

  beforeEach(function() {
    success = spy();
    fail = spy();
    stub(jsonp, "get");
  });

  afterEach(function() {
    jsonp.get.restore();
  });

  describe("for an invalid email", function() {
    beforeEach(function() {
      api.profile(invalidEmail, success, fail);
    });

    it("invokes the 'fail' callback", function() {
      expect(fail.calledOnce).to.be(true);
      expect(fail.calledWithExactly(invalidEmail)).to.be(true);
    });
  });

  describe("for a valid email", function() {
    beforeEach(function() {
      api.profile(email, success, fail);
    });

    it("performs a request to the expected URL", function() {
      const url = `https://secure.gravatar.com/${emailMD5}.json`;
      expect(jsonp.get.calledOnce).to.be(true);
      expect(jsonp.get.withArgs(url).calledOnce).to.be(true);
    });

    describe("with an associated Gravatar profile", function() {
      const response = {entry: [{someAttr: "someAttr"}]};

      beforeEach(function() {
        invokeCallback(null, response);
      });

      it("invokes the 'success' callback with `email` and `img`", function() {
        expect(success.calledOnce).to.be(true);
        expect(success.calledWithExactly(email, response.entry[0])).to.be(true)
      });
    });

    describe("with an associated empty Gravatar profile", function() {
      const response = {entry: []};

      beforeEach(function() {
        invokeCallback(null, response);
      });

      it("invokes the 'success' callback with `email` and `img`", function() {
        expect(fail.calledOnce).to.be(true);
        expect(fail.calledWithExactly(email)).to.be(true)
      });
    });

    describe("without an associated Gravatar profile", function() {
      const error = {code: "someCode"};

      beforeEach(function() {
        invokeCallback(error);
      });

      it("invokes the 'fail' callback with `email`", function() {
        expect(fail.calledOnce).to.be(true);
        expect(fail.calledWithExactly(email, error)).to.be(true)
      });
    });
  });
});
