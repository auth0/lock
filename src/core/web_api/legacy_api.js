import IdTokenVerifier from 'idtoken-verifier';
import auth0 from 'auth0-js';
import CordovaAuth0Plugin from 'auth0-js/plugins/cordova';
import request from 'superagent';
import { normalizeError, loginCallback, normalizeAuthParams, webAuthOverrides } from './helper';
import qs from 'qs';

class Auth0LegacyAPIClient {
  constructor(clientID, domain, opts) {
    this.client = null;
    this.authOpt = null;

    this.domain = domain;
    this.clientID = clientID;
    this.tokenIssuer = (opts.overrides && opts.overrides.__token_issuer) || `https://${domain}/`;

    const default_telemetry = {
      name: 'lock.js',
      version: __VERSION__,
      lib_version: auth0.version
    };

    this.client = new auth0.WebAuth({
      clientID: clientID,
      domain: domain,
      redirectUri: opts.redirectUrl,
      responseMode: opts.responseMode,
      responseType: opts.responseType,
      plugins: [new CordovaAuth0Plugin()],
      overrides: webAuthOverrides(opts.overrides),
      _sendTelemetry: opts._sendTelemetry === false ? false : true,
      _telemetryInfo: opts._telemetryInfo || default_telemetry,
      _disableDeprecationWarnings: true
    });

    this.authOpt = {
      popup: !opts.redirect,
      popupOptions: opts.popupOptions,
      sso: opts.sso,
      nonce: opts.nonce,
      state: opts.state
    };
  }

  logIn(options, authParams, cb) {
    // TODO: for passwordless only, try to clean in auth0.js
    // client._shouldRedirect = redirect || responseType === "code" || !!redirectUrl;
    const f = loginCallback(!this.authOpt.popup, cb);
    const auth0Client = this.client;

    const loginOptions = normalizeAuthParams({ ...options, ...this.authOpt, ...authParams });
    if (!options.username && !options.email) {
      if (this.authOpt.popup) {
        auth0Client.popup.authorize({ ...loginOptions, owp: true }, f);
      } else {
        auth0Client.authorize(loginOptions, f);
      }
    } else if (!this.authOpt.sso && this.authOpt.popup) {
      auth0Client.client.loginWithResourceOwner(loginOptions, f);
    } else if (this.authOpt.popup) {
      auth0Client.popup.loginWithCredentials({ ...loginOptions, owp: true }, f);
    } else {
      auth0Client.redirect.loginWithCredentials(loginOptions, f);
    }
  }

  logout(query) {
    this.client.logout(query);
  }

  signUp(options, cb) {
    const { popup, sso } = this.authOpt;
    const { autoLogin } = options;

    delete options.autoLogin;

    const popupHandler = autoLogin && popup && sso ? this.client.popup.preload() : null;

    this.client.signup(options, (err, result) => cb(err, result, popupHandler));
  }

  resetPassword(options, cb) {
    this.client.changePassword(options, cb);
  }

  startPasswordless(options, cb) {
    this.client.startPasswordless(options, err => cb(normalizeError(err)));
  }

  // for legacy, we should not verify the id_token so we reimplemented it here
  // to avoid adding dirt into auth0.js. At some point we will get rid of this.
  parseHash(hash = '', cb) {
    var parsed_qs = qs.parse(hash.replace(/^#?\/?/, ''));
    var state = this.authOpt.state || parsed_qs.state;

    this.client.transactionManager.getStoredTransaction(state);

    if (parsed_qs.hasOwnProperty('error')) {
      var err = {
        error: parsed_qs.error,
        error_description: parsed_qs.error_description
      };

      if (parsed_qs.state) {
        err.state = parsed_qs.state;
      }

      return cb(err);
    }

    if (
      !parsed_qs.hasOwnProperty('access_token') &&
      !parsed_qs.hasOwnProperty('id_token') &&
      !parsed_qs.hasOwnProperty('refresh_token')
    ) {
      return cb(null, null);
    }

    var prof;

    if (parsed_qs.hasOwnProperty('id_token')) {
      var invalidJwt = function(error) {
        var err = {
          error: 'invalid_token',
          error_description: error
        };
        return err;
      };

      var verifier = new IdTokenVerifier({});
      prof = verifier.decode(parsed_qs.id_token).payload;

      if (prof.aud !== this.clientID) {
        return cb(
          invalidJwt(
            'The clientID configured (' +
              this.clientID +
              ') does not match with the clientID set in the token (' +
              prof.aud +
              ').'
          )
        );
      }

      // iss should be the Auth0 domain (i.e.: https://contoso.auth0.com/)
      if (prof.iss !== this.tokenIssuer) {
        return cb(
          invalidJwt(
            'The domain configured (' +
              this.tokenIssuer +
              ') does not match with the domain set in the token (' +
              prof.iss +
              ').'
          )
        );
      }
    }

    cb(null, {
      accessToken: parsed_qs.access_token,
      idToken: parsed_qs.id_token,
      idTokenPayload: prof,
      refreshToken: parsed_qs.refresh_token,
      state: parsed_qs.state
    });
  }

  getUserInfo(token, callback) {
    return this.client.client.userInfo(token, callback);
  }

  // auth0.js does not supports this endpoint because it is deprecated for oidcConformat clients
  // we implemented it here to provide BC support, we will loose it in lock 11.
  getProfile(token, callback) {
    request.get(`https://${this.domain}/tokeninfo?id_token=${token}`).end(function(err, res) {
      if (err) {
        return callback({
          error: err.message,
          error_description: res.text || res.body
        });
      }

      return callback(null, res.body);
    });
  }

  getSSOData(...args) {
    return this.client.client.getSSOData(...args);
  }

  getUserCountry(cb) {
    return this.client.getUserCountry(cb);
  }
}

export default Auth0LegacyAPIClient;
