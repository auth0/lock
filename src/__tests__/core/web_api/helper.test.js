import { webAuthOverrides, normalizeError } from 'core/web_api/helper';

describe('webAuthOverrides', () => {
  it('should return overrides if any field is compatible with WebAuth', function() {
    expect(
      webAuthOverrides({
        __tenant: 'tenant1',
        __token_issuer: 'issuer1',
        __jwks_uri: 'https://jwks.com'
      })
    ).toMatchSnapshot();
  });

  it('should omit overrides that are not compatible with WebAuth', function() {
    expect(
      webAuthOverrides({
        __tenant: 'tenant1',
        __token_issuer: 'issuer1',
        __jwks_uri: 'https://jwks.com',
        backgroundColor: 'blue'
      })
    ).toMatchSnapshot();
  });

  it('should return null if no fields are compatible with WebAuth', function() {
    expect(webAuthOverrides({ backgroundColor: 'blue' })).toBe(null);
  });
});

describe('normalizeError', () => {
  it('does nothing when there is no error', () => {
    const normalized = normalizeError(undefined);
    expect(normalized).toBe(undefined);
  });
  it('should map access_denied error to invalid_user_password when error.error === access_denied', () => {
    const error = {
      error: 'access_denied',
      description: 'foobar'
    };
    const expectedError = {
      code: 'invalid_user_password',
      error: 'invalid_user_password',
      description: 'foobar'
    };
    const actualError = normalizeError(error);
    expect(actualError).toMatchSnapshot();
  });
  it('should map access_denied error to invalid_user_password when error.code === access_denied', () => {
    const error = {
      code: 'access_denied',
      description: 'foobar'
    };
    const expectedError = {
      code: 'invalid_user_password',
      error: 'invalid_user_password',
      description: 'foobar'
    };
    const actualError = normalizeError(error);
    expect(actualError).toMatchSnapshot();
  });
});
