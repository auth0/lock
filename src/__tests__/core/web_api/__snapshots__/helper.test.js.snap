// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeError should map access_denied error to invalid_user_password when error.code === access_denied 1`] = `
Object {
  "code": "invalid_user_password",
  "description": "foobar",
  "error": "invalid_user_password",
}
`;

exports[`normalizeError should map access_denied error to invalid_user_password when error.error === access_denied 1`] = `
Object {
  "code": "invalid_user_password",
  "description": "foobar",
  "error": "invalid_user_password",
}
`;

exports[`webAuthOverrides should omit overrides that are not compatible with WebAuth 1`] = `
Object {
  "__jwks_uri": "https://jwks.com",
  "__tenant": "tenant1",
  "__token_issuer": "issuer1",
}
`;

exports[`webAuthOverrides should return overrides if any field is compatible with WebAuth 1`] = `
Object {
  "__jwks_uri": "https://jwks.com",
  "__tenant": "tenant1",
  "__token_issuer": "issuer1",
}
`;
